### PR TITLE
fix(editors/cleanup): Fix filter issue with in cleanup plugin

### PR DESCRIPTION
**Pull request by JakobVogelsang**

### CODE REVIEW

**Review by danyill — COMMENTED** [review]
I can't fault it from testing the filters and the way they work with the text filtering -- this looks great.
I thought regenerating the snapshots would probably be sufficient to pass the tests.... except they weren't. 

Now that we are not mutating the `hidden` class, the mechanism we use for select all in `filtered-list.ts` doesn't work correctly, which is:

```ts
  private onCheckAll(): void {
    const select = !this.isAllSelected;
    this.items
      .filter(item => !item.disabled && !item.classList.contains('hidden'))
      .forEach(item => (item.selected = select));
  }
``` 

This means that when we hit the select all checkbox, we select the non-visible items:

(1) If I only have the SV filter 
(2) I hit select all
(3) I have five items selected

image

How should we best resolve this? Add a  `!item.classList.contains('hiddenontypefilter')`?
Doesn't seem tidily scalable.

File: src/editors/cleanup/control-blocks-container.ts [hunk 1]
```diff
@@ -164,7 +165,13 @@ export class CleanupControlBlocks extends LitElement {
   private renderListItem(controlBlock: Element): TemplateResult {
     return html`<mwc-check-list-item
       twoline
-      class="cleanupListItem t${controlBlock.tagName}"
+      class="${classMap({
```
[the remark this answers is not in the document]
Comment (JakobVogelsang) [reply src/editors/cleanup/control-blocks-container.ts:168]: I found it more readable, :) That is why I started to use it.

[the remark this answers is not in the document]
Comment (danyill) [reply src/editors/cleanup/control-blocks-container.ts:168]: > How should we best resolve this? Add a `!item.classList.contains('hiddenontypefilter')`? Doesn't seem tidily scalable.

Should we introduce a property to the filtered list component to allow nominated filter classes? e.g. `additionalFilterClasses="hiddenontypefilter"`?

**Comment by danyill** [comment]
> How should we best resolve this? Add a !item.classList.contains('hiddenontypefilter')?
> Doesn't seem tidily scalable.

Would an acceptable solution be to pass an attribute or property to the filtered list element to identify the additional classes required?

**Comment by JakobVogelsang** [comment]
@danyill I did find a simple solution. We just disable the list item that is hidden. That way, the check all still perfectly works out of the box.

**Review by danyill — APPROVED** [review]
> @danyill I did find a simple solution. We just disable the list item that is hidden. That way, the check all still perfectly works out of the box.

I looked at the code. I'm confident this would work.
If you are happy with then we should merge it, :wink: 

However at risk of ?Du machst einen Berg aus einen Maulwurfshügel?

* We could target [:disabled](https://developer.mozilla.org/en-US/docs/Web/CSS/:disabled) with a CSS selector rather than introduce our own CSS classes for modest benefit.
* This highlights what I see as the general deficiency. We are simply leveraging the existing code that looks at the disabled attribute. We can do that because it is orthogonal to the `hidden` CSS class
* This is fine _if_ we only want to filter lists with _two_ mechanisms that touch an element once (our filters with text filtering only touch the hidden class once, and our button filters only touch the disabled attribute).
* This is not a generic approach and doesn't allow us to use three mechanisms to filter. 
* Here is my example. Let's imagine we name our socks Jeff, Dave, Barney and Mike. We also categorise them: crewcut, ankle, midcalf. Finally we categorise them red, green and blue. Each sock can fit any of these categories. 
    * If we were trying to interact with these in a filtered list, we could text search on whatever fields we wanted (e.g. our sock's name) and mutate the `hidden` class. 
    * We could even create some buttons (in a lovely shade of lavender) and use them to filter socks by colour (by mutating the disabled attribute). 
    * What we could not do is add an additional set of buttons or toggles to select socks by type -- crewcut, ankle. We have run out of "properties" that fit with the way the filtered list component works.
* For this reason, I wonder if passing classes to the filtered list component is a more helpful generic mechanism (as we may want to add filters by more than two properties in future).

**Comment by JakobVogelsang** [comment]
All of what you said is very true. I think a good time to rethink its behavior is `open-scd-core`. With filtered-list in `open-scd` we now know what worked well and what could be improved. Once we start migrating it to `open-scd-core` we can tackle the missing feature wishes. I therefore would be very happy about an issue in `open-scd-core` that in detail describes what we expect from this web component in the future. 
@danyill what you say you want to give it a shot and I add stuff, if necessary?

**Pull request merged**